### PR TITLE
add ROC AUC metric

### PR DESCRIPTION
**Pull request by SuvarshaChennareddy**
Hello everyone!
I implemented the ROC AUC score metric that was recommended in the issue #3074. 
Here, the multiclass AUC was implemented using a generalization by Hand and Till in the year 2001 (source: [A Simple Generalisation of the Area Under the ROC Curve for Multiple Class Classification Problems](https://link.springer.com/article/10.1023/A:1010920819831))
@srimadhan11 had already attempted to implement his own version, however I noticed that he was inactive for the past couple of weeks. Hence, I decided to implement it myself. Hope that's all right with everyone.
Any comments on this would be greatly appreciated.
Thank You!

Edit:
Forgot to add this...
Google Colab notebook for the tests: [link](https://colab.research.google.com/drive/1wvKRcGIpRHf4-MRfTm4B5_WnL0aI7kNH?usp=sharing)

### CODE REVIEW

**Comment by rcurtin** [comment]
Hey @SuvarshaChennareddy, thanks for the contribution.  But #3086 is already open---that one has already been partially reviewed, so personally I think we should go with that.  Maybe you could help review that one and see if there are any issues?

**Comment by SuvarshaChennareddy** [comment]
Hello @rcurtin ! First of all thank you for reaching out!

I’ve seen  PR #3086 already and one other thing that made me implement this (other than the 2 week inactivity and the multi-class ROC AUC) was the test case. I calculated the ROC AUC with the scikit-learn machine learning library in python, and the result was approx. 0.70833. (Sri Madhan's result was approx. 0.6667). However I wasn’t completely sure whether @srimadhan11 was wrong. Hence I went ahead and implemented this and I too got approx. 0.70833. (I used the property that the ROC AUC remains the same for any 2 sequences of predictions that have the same relative order with respect to the same labels. I’ve included the link to the google Colab notebook that shows this in the previous comment)

Would you like me to address this issue in #3086 ?

If required, I’ll close this PR. I'm sorry for acting so quickly. May I make a separate PR in the future for multiclass AUC (if this PR needs to be closed)?

**Comment by rcurtin** [comment]
Well, I'd prefer to only merge one of these versions (having two confuses users).  But I don't want either you or @srimadhan11 to feel like either of your works are unappreciated.  In your response I can see two issues you've pointed out:

 * *The implementation gives different results than scikit.*  This does not necessarily mean that the implementation in #3086 is wrong---scikit is not the authority on all things.  Regardless of the implementation we go with, we should determine what the cause of the difference is---it may be a bug in the implementation, or it may be a difference in interpretation of what the AUC actually is.  (I suspect it is actually the choice to use the trapezoid rule to compute AUC, and there's a comment open about that already.)

 * *The implementation in #3086 doesn't support multi-class AUC*.  To me it seems like we should try to generalize the implementation that is already there in some graceful way, either as a follow-up PR or as a part of #3086 itself.

People do get busy, so it may be a little while until we manage to address all the comments in #3086, but if your goal is overall just to contribute to the library in short order (instead of waiting to see that one through), there are tons of other open issues too. :)

**Comment by SuvarshaChennareddy** [comment]
@rcurtin Hello and thanks for the quick reply!

I get that sklearn is not the deciding factor, that's why I went ahead and tried implementing this. It started out as an activity to get to know the code base better and how mlpack works. 
As for the mult-iclass AUC, I too think it's a great idea to work on that later.
I think I might have acted a bit rash. Time is not really a problem for me (I think I might have gotten a bit excited since it's my first time actually contributing to a huge codebase...sorry). I'll close this one and try to help review the PR #3086. Maybe in the near future, I can help out with the muli-class AUC implementation? Does that seem like a good idea?
